Fix to return `initial value` for SVG Length in case of parser error and being invalid

Ahmad-S792 opened the pull request:
#### 38b65cc65f133e6e5ac0afbf9a24149897e556b1
<pre>
Fix to return `initial value` for SVG Length in case of parser error and being invalid
<a href="https://bugs.webkit.org/show_bug.cgi?id=279379">https://bugs.webkit.org/show_bug.cgi?id=279379</a>
<a href="https://rdar.apple.com/136102554">rdar://136102554</a>

Reviewed by Darin Adler.

Our current implementation for initial values was not taking account when
there was any parser error or when the value was empty string or null, this
patch fixes the bug by ensuring that we return the initial value as per
web-standard in those cases.

NOTE: We still fail test about `fx` and `fy`, where the initial value fallback
is unclear in specification, so this patch does not fix those cases.

* Source/WebCore/svg/SVGFilterElement.cpp:
(WebCore::SVGFilterElement::attributeChanged):
* Source/WebCore/svg/SVGFilterPrimitiveStandardAttributes.cpp:
(WebCore::SVGFilterPrimitiveStandardAttributes::attributeChanged):
* Source/WebCore/svg/SVGLengthList.cpp:
(WebCore::SVGLengthList::parse):
* Source/WebCore/svg/SVGLengthValue.cpp:
(WebCore::SVGLengthValue::construct):
* Source/WebCore/svg/SVGLengthValue.h:
(WebCore::SVGLengthValue::construct):
* Source/WebCore/svg/SVGLinearGradientElement.cpp:
(WebCore::SVGLinearGradientElement::attributeChanged):
* Source/WebCore/svg/SVGMarkerElement.cpp:
(WebCore::SVGMarkerElement::attributeChanged):
* Source/WebCore/svg/SVGMaskElement.cpp:
(WebCore::SVGMaskElement::attributeChanged):
* Source/WebCore/svg/SVGRadialGradientElement.cpp:
(WebCore::SVGRadialGradientElement::attributeChanged):
* Source/WebCore/svg/SVGSVGElement.cpp:
(WebCore::SVGSVGElement::attributeChanged):

* imported/w3c/web-platform-tests/svg/types/scripted/SVGAnimatedLength-initial-values-expected.txt: Progressions

Canonical link: <a href="https://commits.webkit.org/303577@main">https://commits.webkit.org/303577@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/187e6588070eb4207fd4d58556489cd72cae126f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132821 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5322 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43906 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140353 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/84849 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d9caa900-5723-4c0e-a941-faf417969012) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/134691 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/5655 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5185 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101556 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/68857 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/c10158c1-ea86-4825-a545-9f4c4575f7a1) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135767 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/3969 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118996 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82349 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/cb9ee19b-2e14-44c5-8b7e-5f5cd469bbbe) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/3856 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/1542 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/83586 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/112844 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37114 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143008 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4991 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37699 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109930 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5073 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4294 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110108 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27924 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3816 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115262 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/58506 "Built successfully") | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5045 "Build is in progress. Recent messages:OS: Tahoe (26.0), Xcode: 26.0.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings)") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/33615 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4883 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68497 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5135 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5003 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/svg/SVGRadialGradientElement.h:
```diff
@@ -67,8 +67,8 @@ class SVGRadialGradientElement final : public SVGGradientElement {
-    Ref<SVGAnimatedLength> m_fx { SVGAnimatedLength::create(this, SVGLengthMode::Width) };
-    Ref<SVGAnimatedLength> m_fy { SVGAnimatedLength::create(this, SVGLengthMode::Height) };
+    Ref<SVGAnimatedLength> m_fx { SVGAnimatedLength::create(this, SVGLengthMode::Width, "50%"_s) };
+    Ref<SVGAnimatedLength> m_fy { SVGAnimatedLength::create(this, SVGLengthMode::Height, "50%"_s) };
```
I think the specs do not say the initial values of `fx` and `fy` are 50% and 50%. Where did you get these initial values from?

Ahmad-S792 replied:
https://svgwg.org/svg2-draft/pservers.html#RadialGradientElementFXAttribute

When it is not specified, it coincides with `cx` and same for `fy`.

shallawa replied:
This is the only thing we do in `SVGRadialGradientElement::collectGradientAttributes()` to respect this part of the specs:

```
    // Handle default values for fx/fy
    if (!attributes.hasFx())
        attributes.setFx(attributes.cx());

    if (!attributes.hasFy())
        attributes.setFy(attributes.cy());
```

But I do not see that we change the fx and fy properties every time the fx, fy, cx are cy attributes change. For example:

1. If the fx attribute is removed, we should set the fx property to the value of the cx property.
2. If cx attribute changes and fx is not specified, the new value of cx is set to fx.

Ahmad-S792 replied:
@karlcow - Can we address this in specification or if you can shed some night.

karlcow replied:
https://svgwg.org/svg2-draft/pservers.html#RadialGradientElementFXAttribute

> [initial value](https://svgwg.org/svg2-draft/types.html#TermInitialValue)
see below
> [Animatable](https://svgwg.org/specs/animations/#Animatable)
yes

Let's look below. (The spec would deserve example each time there is this kind of prose.)

> 1. If attribute ‘[fx](https://svgwg.org/svg2-draft/pservers.html#RadialGradientElementFXAttribute)’ is not specified, ‘[fx](https://svgwg.org/svg2-draft/pservers.html#RadialGradientElementFXAttribute)’ will coincide with the presentational value of ‘[cx](https://svgwg.org/svg2-draft/pservers.html#RadialGradientElementCXAttribute)’ for the element whether the value for 'cx' was inherited or not. 
> 2. If the element references an element that specifies a value for 'fx', then the value of 'fx' is inherited from the referenced element.

* if `cx="40"` and no `fx` is in the markup, then `fx` value is `40` (`cx` is not necessarily defined here and can be inherited too. **Though I wonder how you inherit the value of `cx`? What would be the markup for this case**)
* the case in 2 is for `<defs>` I guess. This could be clearer in the spec too. 

So the initial value of `fx` when 
* it is not specified, aka `<radialGradient cx="40" cy="40" r="40">…`,  is the current value of `cx`, aka 40.
* of if the value is empty, aka `<radialGradient cx="40" cy="40" r="40" fx="">…` is the current value of `cx` aka 40.

if the value of `cx` was not specified then yes it would be "50%" aka `<radialGradient cy="40" r="40">…` 

See my test cases in https://codepen.io/webcompat/full/GgZQMgZ
<img width="2646" height="1251" alt="Screenshot 2025-11-26 at 10 27 25" src="https://github.com/user-attachments/assets/ffbdbd21-6060-41a3-bf75-6dfac240388f" />

right now, interop is failing for the case: **`cx=40 / fx=""`**


```
    <svg viewBox="0 0 80 80">
      <defs>
        <radialGradient id="MyGradient11" gradientUnits="userSpaceOnUse" cx="40" fx="" cy="40" r="40">
          <stop offset="0%" stop-color="#A8F" />
          <stop offset="50%" stop-color="#FDC" />
          <stop offset="100%" stop-color="#A8F" />
        </radialGradient>
      </defs>

      <!-- The rectangle is filled using a radial gradient paint server -->
      <rect fill="url(#MyGradient11)" stroke="black" stroke-width="2" x="0" y="0" width="80" height="80" />
    </svg>
```

karlcow replied:
And **`cx=40 / fx=80ABC`**

```
<svg viewBox="0 0 80 80">
      <defs>
        <radialGradient id="MyGradient9" gradientUnits="userSpaceOnUse" cx="40" fx="80ABC" cy="40" r="40">
          <stop offset="0%" stop-color="#A8F" />
          <stop offset="50%" stop-color="#FDC" />
          <stop offset="100%" stop-color="#A8F" />
        </radialGradient>
      </defs>

      <!-- The rectangle is filled using a radial gradient paint server -->
      <rect fill="url(#MyGradient9)" stroke="black" stroke-width="2" x="0" y="0" width="80" height="80" />
    </svg>
```

Ahmad-S792 replied:
@karlcow - I undid the change, we can explore it in future but just to flag adding this does not regress your test case.

karlcow replied:
I created a test to observer the behavior of browsers. 

with regards to what @shallawa says:

> 1. If the fx attribute is removed, we should set the fx property to the value of the cx property.

Both Firefox and Chrome do not respect this. 
They set the value to 50%. They do not follow the spec. 

> 2. If cx attribute changes and fx is not specified, the new value of cx is set to fx.

Both Firefox and Chrome do not respect this. 
They set the value to 50%. They do not follow the spec. 


I think the spec needs to change. @Ahmad-S792 opened an issue.
https://github.com/w3c/svgwg/issues/1054
I put in on the agenda for this week. 

The current WPT tests do not showcase this part of the spec. 

See the test I uploaded in 
https://github.com/user-attachments/files/24730583/gradient-test.html

So the fix to default to 50% brings WebKit closer to the actual other browsers behavior.

Depending on the outcome of the discussion this week in the SVGWG.
We could  fix the dependency of fx/cx, fy/cy if necessary. Or just ignore if the spec is changing. 

In the meantime I think we should merge this, with a default value of 50%. because it is basically what browsers are doing.

shallawa commented on Source/WebCore/svg/SVGFilterElement.cpp:
```diff
@@ -81,18 +82,34 @@ void SVGFilterElement::attributeChanged(const QualifiedName& name, const AtomStr
+        auto length = SVGLengthValue::construct(SVGLengthMode::Width, newValue, parseError);
+        if (parseError != NoError || newValue.isNull())
+            length = SVGLengthValue(SVGLengthMode::Width, "-10%"_s);
+        Ref { m_x }->setBaseValInternal(length);
```
This code is repeated multiple times in this PR. It parses an `SVGLengthValue` if it is invalid it sets the baseVal of the `SVGAnimatedProperty` to the initial value. Should not this code be put in one function so this change is smaller?

Ahmad-S792 replied:
It is something currently done as well, so I opted for this. I am OK with using helper function or reduce repetition but I was thinking of doing it as future follow-up.

shallawa replied:
I think using table / data driven as Sam suggested or a single function to set the properties should be a better approach. Adding repeated code in multiple places and then trying to shorten it later does not seem an efficient way to do things.

darinadler replied:
I agree that we should include the helper function in the initial patch. The pattern is simple enough that I expect it can be done cleanly.

karlcow replied:
@Ahmad-S792 

in `SVGParserUtilities.h`

```
void parseAndSetLengthValue(Ref<SVGAnimatedLength>& animatedLength, SVGLengthMode mode, const AtomString& newValue, const String& fallbackValue);
```

in `SVGParserUtilities.cpp`

```
void parseAndSetLengthValue(Ref<SVGAnimatedLength>& animatedLength, SVGLengthMode mode, const AtomString& newValue, const String& fallbackValue);
{
    SVGParsingError parseError = NoError;
    auto length = SVGLengthValue::construct(mode, newValue, parseError);
    if (parseError != NoError || newValue.isNull())
        length = SVGLengthValue(mode, fallbackValue);
    Ref { animatedLength }->setBaseValInternal(length);
}
```


So you can rewrite with:

```
case AttributeNames::xAttr: {
    SVGParserUtilities::parseAndSetLengthValue(m_x, SVGLengthMode::Width, newValue, "-10%"_s);
    break;
}
case AttributeNames::yAttr: {
    SVGParserUtilities::parseAndSetLengthValue(m_y, SVGLengthMode::Height, newValue, "-10%"_s);
    break;
}
case AttributeNames::widthAttr: {
    SVGParserUtilities::parseAndSetLengthValue(m_width, SVGLengthMode::Width, newValue, "120%"_s);
    break;
}
case AttributeNames::heightAttr: {
    SVGParserUtilities::parseAndSetLengthValue(m_height, SVGLengthMode::Height, newValue, "120%"_s);
    break;
}
```

brentfulgham replied:
I agree with @darinadler and @karlcow ! This will be much easier to read and reason about.

darinadler commented on LayoutTests/svg/dom/svganimatedlength-initial-values-expected.txt:
```diff
@@ -0,0 +1,197 @@
+CONSOLE MESSAGE: Error: Invalid value for <circle> attribute cx="foobar"
```
Is it good to log these in the console? I guess maybe it's helpful for website developers?

darinadler commented on Source/WebCore/svg/SVGFilterPrimitiveStandardAttributes.cpp:
```diff
@@ -53,18 +53,34 @@ void SVGFilterPrimitiveStandardAttributes::attributeChanged(const QualifiedName&
+        auto length = SVGLengthValue::construct(SVGLengthMode::Width, newValue, parseError);
+        if (parseError != NoError || newValue.isNull())
+            length = SVGLengthValue(SVGLengthMode::Width, "0%"_s);
+        Ref { m_x }->setBaseValInternal(length);
```
```suggestion
        parse(m_x, SVGLengthMode::Width, "0%"_s);
```
The `parse` function should be easy to write.

karlcow replied:
Should the function be in SVGParserUtilities.cpp

karlcow replied:
see https://github.com/WebKit/WebKit/pull/38414/files#r2106870367

brentfulgham reviewed (changes requested):
Please revise your patch to create and adopt a 'parse' function as @karlcow and @darinadler suggested.

darinadler commented on Source/WebCore/svg/SVGLengthValue.h:
```diff
@@ -71,6 +71,7 @@ class SVGLengthValue {
 
     static std::optional<SVGLengthValue> construct(SVGLengthMode, StringView);
     static SVGLengthValue construct(SVGLengthMode, StringView, SVGParsingError&, SVGLengthNegativeValuesMode = SVGLengthNegativeValuesMode::Allow);
+    static SVGLengthValue constructLengthValueWithFallback(SVGLengthMode, const AtomString&, const String&);
```
```suggestion
    static SVGLengthValue construct(SVGLengthMode, StringView value, ASCIILiteral fallback);
```
This is a member of the `SVGLengthValue` class and so we don't need to repeat words from the name of the class here. Also, this should take a `StringView` since that's what the constructor takes and for the fallback I suggest taking `ASCIILiteral` so we don't create and destroy a String when we are not falling back. Without argument names it's not clear what the two different strings are. I think we can name this without the words "with fallback" because the overloading makes it unambiguous enough.

Ahmad-S792 replied:
Refactored and addressed this feedback. :-)

darinadler commented on Source/WebCore/svg/SVGFilterElement.cpp:
```diff
@@ -82,18 +82,26 @@ void SVGFilterElement::attributeChanged(const QualifiedName& name, const AtomStr
     }
-    case AttributeNames::xAttr:
-        Ref { m_x }->setBaseValInternal(SVGLengthValue::construct(SVGLengthMode::Width, newValue, parseError));
+    case AttributeNames::xAttr: {
+        auto length = SVGLengthValue::constructLengthValueWithFallback(SVGLengthMode::Width, newValue, "-10%"_s);
+        Ref { m_x }->setBaseValInternal(length);
         break;
```
```suggestion
    case AttributeNames::xAttr:
        Ref { m_x }->setBaseValInternal(SVGLengthValue::construct(SVGLengthMode::Width, newValue, "-10%"_s));
```
It's not an improvement to add a local variable, we shouldn't do that.

Ahmad-S792 replied:
Refactored and addressed this feedback. :-)

darinadler commented on Source/WebCore/svg/SVGLengthValue.h:
```diff
@@ -71,6 +71,7 @@ class SVGLengthValue {
 
     static std::optional<SVGLengthValue> construct(SVGLengthMode, StringView);
     static SVGLengthValue construct(SVGLengthMode, StringView, SVGParsingError&, SVGLengthNegativeValuesMode = SVGLengthNegativeValuesMode::Allow);
+    static SVGLengthValue construct(SVGLengthMode, StringView, ASCIILiteral);
```
```suggestion
    static SVGLengthValue construct(SVGLengthMode, StringView, ASCIILiteral, SVGLengthNegativeValuesMode = SVGLengthNegativeValuesMode::Allow);
```

darinadler commented on Source/WebCore/svg/SVGRadialGradientElement.cpp:
```diff
@@ -72,23 +72,27 @@ void SVGRadialGradientElement::attributeChanged(const QualifiedName& name, const
+            length = SVGLengthValue(SVGLengthMode::Other, "0%"_s);
+        Ref { m_fr }->setBaseValInternal(length);
         break;
+    }
```
```suggestion
    case AttributeNames:frAttr:
        Ref { m_fr }->setBaseValInternal(SVGLengthValue::construct(SVGLengthMode::Other, newValue, "0%"_s, SVGLengthNegativeValuesMode::Forbid);
        break;
```

darinadler reviewed (changes requested):
I think we need to resolve the comment from @shallawa before landing this.

darinadler commented on Source/WebCore/svg/SVGLengthValue.h:
```diff
@@ -67,8 +67,7 @@ class SVGLengthValue {
 
-    static std::optional<SVGLengthValue> construct(SVGLengthMode, StringView);
-    static SVGLengthValue construct(SVGLengthMode, StringView, SVGParsingError&, SVGLengthNegativeValuesMode = SVGLengthNegativeValuesMode::Allow);
+    static SVGLengthValue construct(SVGLengthMode, StringView, SVGParsingError&, SVGLengthNegativeValuesMode = SVGLengthNegativeValuesMode::Allow, std::optional<ASCIILiteral> fallbackValue = std::nullopt);
```
`ASCIILiteral` has a null value, so `fallbackValue` can just be an `ASCIILiteral` rather than a `std::optional<ASCIILiteral>`.

Not new to this patch, to consider for future improvement: the name `construct` is not a typical name for a parsing function. It's more important that it parses than the fact that the value it returns is "constructed". I think this should eventually be renamed `parse` and should be changed to return `std::tuple<SVGLengthValue, SVGParserError>` rather than using an out argument. We would refactor the call sites a bit.

Ahmad-S792 replied:
@darinadler - Thanks for review, I updated now and accommodated your feedback of `ASCIILiteral` instead of `std::optional<ASCIILiteral>`. As for future work, I will file follow-up bug to see, if it is something I can tackle myself but thanks a lot for help. I also undid Said's concern on other change and I am in discussion with Karl on how to proceed, so I think it is now ready to land. After green EWS, I will land it. :-)

Ahmad-S792 replied:
https://bugs.webkit.org/show_bug.cgi?id=303150 <- for future work.

Ahmad-S792 committed:
Fix to return `initial value` for SVG Length in case of parser error and being invalid
https://bugs.webkit.org/show_bug.cgi?id=279379
rdar://136102554

Reviewed by Darin Adler.

Our current implementation for initial values was not taking account when
there was any parser error or when the value was empty string or null, this
patch fixes the bug by ensuring that we return the initial value as per
web-standard in those cases.

NOTE: We still fail test about `fx` and `fy`, where the initial value fallback
is unclear in specification, so this patch does not fix those cases.

* Source/WebCore/svg/SVGFilterElement.cpp:
(WebCore::SVGFilterElement::attributeChanged):
* Source/WebCore/svg/SVGFilterPrimitiveStandardAttributes.cpp:
(WebCore::SVGFilterPrimitiveStandardAttributes::attributeChanged):
* Source/WebCore/svg/SVGLengthList.cpp:
(WebCore::SVGLengthList::parse):
* Source/WebCore/svg/SVGLengthValue.cpp:
(WebCore::SVGLengthValue::construct):
* Source/WebCore/svg/SVGLengthValue.h:
(WebCore::SVGLengthValue::construct):
* Source/WebCore/svg/SVGLinearGradientElement.cpp:
(WebCore::SVGLinearGradientElement::attributeChanged):
* Source/WebCore/svg/SVGMarkerElement.cpp:
(WebCore::SVGMarkerElement::attributeChanged):
* Source/WebCore/svg/SVGMaskElement.cpp:
(WebCore::SVGMaskElement::attributeChanged):
* Source/WebCore/svg/SVGRadialGradientElement.cpp:
(WebCore::SVGRadialGradientElement::attributeChanged):
* Source/WebCore/svg/SVGSVGElement.cpp:
(WebCore::SVGSVGElement::attributeChanged):

* imported/w3c/web-platform-tests/svg/types/scripted/SVGAnimatedLength-initial-values-expected.txt: Progressions

Canonical link: https://commits.webkit.org/303577@main
